com.unity.multiplayer.mlapi/Runtime/Core/SnapshotSystem.cs

```diff
@@ -104,29 +116,166 @@ public int AddEntry(ulong networkObjectId, int behaviourIndex, int variableIndex
+        }
+
         /// <summary>
         /// Allocate memory from the buffer for the Entry and update it to point to the right location
         /// </summary>
         /// <param name="entry">The entry to allocate for</param>
         /// <param name="size">The need size in bytes</param>
-        public void AllocateEntry(ref Entry entry, long size)
+        public void AllocateEntry(ref Entry entry, int index, int size)
         {
             // todo --M1--
             // this will change once we start reusing the snapshot buffer memory
             // todo: deal with free space
             // todo: deal with full buffer
 
-            entry.Position = (ushort)FreeMemoryPosition;
+            int pos;
+
+            if (entry.Length > 0)
+            {
+                Allocator.Deallocate(index);
+            }
+
+            bool ret = Allocator.Allocate(index, size, out pos);
+
+            if (!ret)
+            {
+                //todo: error handling
+            }
+
+            entry.Position = (ushort)pos;
             entry.Length = (ushort)size;
-            FreeMemoryPosition += (int)size;
+        }
+
+        /// <summary>
+        /// Read the buffer part of a snapshot
+        /// Must match WriteBuffer
+        /// The stream is actually a memory stream and we seek to each variable position as we deserialize them
+        /// </summary>
+        /// <param name="reader">The NetworkReader to read our buffer of variables from</param>
+        /// <param name="snapshotStream">The stream to read our buffer of variables from</param>
+        internal void ReadBuffer(NetworkReader reader, Stream snapshotStream)
+        {
+            int snapshotSize = reader.ReadUInt16();
+
+            snapshotStream.Read(Buffer, 0, snapshotSize);
+
+            for (var i = 0; i < LastEntry; i++)
+            {
+                if (Entries[i].Fresh && Entries[i].Key.TickWritten > 0)
+                {
+                    var networkVariable = FindNetworkVar(Entries[i].Key);
+
+                    if (networkVariable != null)
+                    {
+                        Stream.Seek(Entries[i].Position, SeekOrigin.Begin);
+
+                        // todo: consider refactoring out in its own function to accomodate
+                        // other ways to (de)serialize
+                        // todo --M1--
+                        // Review whether tick still belong in netvar or in the snapshot table.
+                        networkVariable.ReadDelta(Stream, m_NetworkManager.IsServer);
+                    }
+                }
+
+                Entries[i].Fresh = false;
+            }
+        }
+
+        /// <summary>
+        /// Read the snapshot index from a buffer
+        /// Stores the entry. Allocates memory if needed. The actual buffer will be read later
+        /// </summary>
+        /// <param name="reader">The reader to read the index from</param>
+        internal void ReadIndex(NetworkReader reader)
```

Comment:
I had to stop and think about the meaning of `ReadIndex`.  It is it not 'ReadNextEntries'?